src/main/java/in/projecteka/consentmanager/common/CentralRegistry.java

```diff
@@ -25,8 +26,31 @@
     }
 
     public Mono<String> authenticate() {
-        return clientRegistryClient
-                .getTokenFor(properties.getClientId(), properties.getXAuthToken())
-                .map(session -> format("%s %s", session.getTokenType(), session.getAccessToken()));
+        var accessToken = accessTokenCache.getIfPresent("accessToken").switchIfEmpty(Mono.just(""));
```

Comment:
```
Mono<String> tokenUsingSecret =
				clientRegistryClient.getTokenFor(properties.getClientId(), properties.getXAuthToken())
						.flatMap(session -> accessTokenCache.put("accessToken", session.getAccessToken())
								.then(refreshTokenCache.put("refreshToken", session.getRefreshToken()))
								.thenReturn(session.getAccessToken()));
		Mono<String> tokenUsingRefreshToken = refreshTokenCache.getIfPresent("refreshToken")
				.switchIfEmpty(Mono.defer(() ->tokenUsingSecret))
				.flatMap(token -> clientRegistryClient.getTokenWithRefreshToken(properties.getClientId(),
						properties.getXAuthToken(),
						token)
						.flatMap(session -> accessTokenCache.put("accessToken", session.getAccessToken())
								.then(refreshTokenCache.put("refreshToken", session.getRefreshToken()))
								.thenReturn(session.getAccessToken())));
		return accessTokenCache.getIfPresent("accessToken")
				// but ideally it should be like (if empty call tokenUsingSecret, if expired call tokenUsingRefreshToken)
				.switchIfEmpty(Mono.defer(() -> tokenUsingRefreshToken))
				.map(token -> String.format("%s %s", "Bearer", token));
```

it should be like (if empty call tokenUsingSecret, if expired call tokenUsingRefreshToken)

Reply:
We tried with switchIfEmpty But It's not working properly.

Reply:
and here, what we are doing is, if the access token expires it'll be removed from the cache. Hence if the access token is empty and refreshToken is there then we are calling get token  using refreshToken

Reply:
If the both the tokens expires, that means both will be removed from the Cache and hence making a call to get the fresh token

Reply:
> We tried with switchIfEmpty But It's not working properly.

Does not seem to be right answer. Are you sure even my code is not working? if not, debug it.

Since Keycloak does not expect you to put send the access\_token when passing refresh token, whatever you have done so far is fine. make sure my refactoring works fine and also followed my advise about the caching part.

Reply:
I'm not saying that your code won't work.

Can we have quick call for better understanding?.

Reply:
and I've tried your implementation (which was also our previous implementation ) but for the very first It's giving error

Reply:
Let's get on a call tomorrow then.